libkbfs/folder_block_ops.go

```diff
@@ -2587,16 +2604,97 @@ func (fbo *folderBlockOps) SearchForNodes(ctx context.Context,
 			md.data.Dir.BlockPointer, rootPath)
 	}
 
-	_, err := fbo.searchForNodesInDirLocked(ctx, lState, cache, newPtrs, md, rootPath,
-		nodeMap, numNodesFound)
+	_, err := fbo.searchForNodesInDirLocked(ctx, lState, cache, newPtrs, md,
+		node, rootPath, nodeMap, numNodesFound)
 	if err != nil {
 		return nil, err
 	}
 
-	// Return the whole map even if some nodes weren't found.
+	if rootPtr != cache.PathFromNode(node).tailPointer() {
+		return nil, searchWithOutOfDateCacheError{}
+	}
+
 	return nodeMap, nil
 }
 
+func (fbo *folderBlockOps) searchForNodesLocked(ctx context.Context,
+	lState *lockState, cache NodeCache, ptrs []BlockPointer,
+	newPtrs map[BlockPointer]bool, md ReadOnlyRootMetadata) (
+	map[BlockPointer]Node, NodeCache, error) {
+	fbo.blockLock.AssertAnyLocked(lState)
+
+	// First try the passed-in cache.  If it doesn't work because the
+	// cache is out of date, try again with a clean cache.
+	nodeMap, err := fbo.trySearchWithCacheLocked(ctx, lState, cache, ptrs,
+		newPtrs, md)
+	if _, ok := err.(searchWithOutOfDateCacheError); ok {
+		// The md is out-of-date, so use a throwaway cache so we
+		// don't pollute the real node cache with stale nodes.
+		fbo.log.CDebugf(ctx, "Root node %v doesn't exist in the node "+
+			"cache; using a throwaway node cache instead",
+			md.data.Dir.BlockPointer)
+		cache = newNodeCacheStandard(fbo.folderBranch)
+		nodeMap, err = fbo.trySearchWithCacheLocked(ctx, lState, cache, ptrs,
```

Comment:
Because of the recursive calls to `searchForNodesInDirLocked`, there are multiple potential unlocks as part of the path tree traversal. I don't have a solid mental model of how this works yet, but I suspect that even a clean cache might get invalidated by writes while we're traversing for search. WDYT?